# Предобработка данных о продажах компьютерных игр

- Автор: Мелания Ланская
- Дата: 04.06.2026

### Цели и задачи проекта

В данном проекте используются данные о продажах компьютерных игр на различных платформах. Цель работы - изучить структуру данных, проверить их качество и корректность, выполнить предобработку и подготовить датасет для дальнейшего анализа, получив необходимый срез данных.

### Описание данных

Датасет `new_games.csv` содержит информацию о продажах игр разных жанров и платформ, а также пользовательские и экспертные оценки игр.

Описание столбцов:
- `Name` — название игры.
- `Platform` — название платформы.
- `Year of Release` — год выпуска игры.
- `Genre` — жанр игры.
- `NA sales` — продажи в Северной Америке (млн копий).
- `EU sales` — продажи в Европе (млн копий).
- `JP sales` — продажи в Японии (млн копий).
- `Other sales` — продажи в остальных регионах (млн копий).
- `Critic Score` — оценка критиков (от 0 до 100).
- `User Score` — оценка пользователей (от 0 до 10).
- `Rating` — возрастной рейтинг ESRB.

### Содержимое проекта

Исследование будет выполнено в несколько этапов:
1. Загрузка данных и знакомство с ними.
2. Проверка ошибок в данных и их предобработка.
3. Фильтрация данных.
4. Категоризация данных.
5. Формулирование итоговых выводов.

---

## 1. Загрузка данных и знакомство с ними

На первом этапе исследования необходимо загрузить данные и ознакомиться с их структурой. Для этого загрузим необходимые библиотеки для анализа данных и данные из датасета `new_games.csv`.


In [53]:
# Импортируем библиотеку pandas
import pandas as pd

In [54]:
# Выгружаем данные из датасета new_games.csv в датафрейм games
games = pd.read_csv('...new_games.csv')

Затем выведем основную информацию о данных с помощью метода `info()` и первые строки датафрейма.


In [55]:
# Выводим информацию о датафрейме
games.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16956 entries, 0 to 16955
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Name             16954 non-null  object 
 1   Platform         16956 non-null  object 
 2   Year of Release  16681 non-null  float64
 3   Genre            16954 non-null  object 
 4   NA sales         16956 non-null  float64
 5   EU sales         16956 non-null  object 
 6   JP sales         16956 non-null  object 
 7   Other sales      16956 non-null  float64
 8   Critic Score     8242 non-null   float64
 9   User Score       10152 non-null  object 
 10  Rating           10085 non-null  object 
dtypes: float64(4), object(7)
memory usage: 1.4+ MB


In [56]:
# Выводим первые строки датафрейма на экран
games.head()

,Name,Platform,Year of Release,Genre,NA sales,EU sales,JP sales,Other sales,Critic Score,User Score,Rating
0,Wii Sports,Wii,2006.0,Sports,41.36,28.96,3.77,8.45,76.0,8,E
1,Super Mario Bros.,NES,1985.0,Platform,29.08,3.58,6.81,0.77,NaN,NaN,NaN
2,Mario Kart Wii,Wii,2008.0,Racing,15.68,12.76,3.79,3.29,82.0,8.3,E
3,Wii Sports Resort,Wii,2009.0,Sports,15.61,10.93,3.28,2.95,80.0,8,E
4,Pokemon Red/Pokemon Blue,GB,1996.0,Role-Playing,11.27,8.89,10.22,1.00,NaN,NaN,NaN


In [57]:
# Определяем размер датафрейма
games.shape

(16956, 11)

Датасет `new_games.csv` содержит 11 столбцов и 16956 строк, в которых представлена информация о компьютерных играх, их жанрах, платформах, продажах в различных регионах мира, а также пользовательских и экспертных оценках.

Изучим типы данных и их корректность:

- **Числовые значения с плавающей запятой (`float64`).** Четыре столбца представлены типом `float64`: `Year of Release`, `NA sales`, `Other sales`, `Critic Score`. Для столбцов с продажами и оценками такой тип данных является корректным, поскольку значения могут содержать дробную часть. Столбец `Year of Release` содержит годы выпуска игр и в дальнейшем может быть преобразован в целочисленный тип после обработки пропусков.

- **Строковые данные (`object`).** Семь столбцов имеют тип данных `object`: `Name`, `Platform`, `Genre`, `EU sales`, `JP sales`, `User Score` и `Rating`.
    - Столбцы `Name`, `Platform`, `Genre` и `Rating` содержат текстовую информацию, поэтому тип `object` для них является корректным.
    - Столбцы `EU sales` и `JP sales` содержат данные о продажах игр, однако имеют тип `object`, хотя должны быть представлены числовыми значениями. Вероятно, в этих столбцах присутствуют некорректные записи или строковые значения, которые необходимо дополнительно проверить.
    - Столбец `User Score` содержит пользовательские оценки и также должен иметь числовой тип данных. Причиной использования типа `object` может быть наличие строковых значений, например `unknown`.

После анализа структуры данных можно сделать вывод, что часть столбцов требует дополнительной обработки и преобразования типов данных.

Также по результатам метода `info()` видно наличие пропущенных значений в нескольких столбцах датафрейма. Особенно заметны пропуски в столбцах, содержащих пользовательские и экспертные оценки, а также возрастной рейтинг ESRB. Более подробно причины возникновения пропусков и способы их обработки будут рассмотрены на этапе предобработки данных.

Кроме того, названия столбцов содержат пробелы и используют смешанный регистр символов. Для удобства дальнейшей работы их необходимо привести к стилю `snake_case`.

---

## 2.  Проверка ошибок в данных и их предобработка


### 2.1. Названия, или метки, столбцов датафрейма

Перед началом предобработки необходимо проверить названия столбцов датафрейма.

In [58]:
# Выводим названия всех столбцов датафрейма
games.columns

Index(['Name', 'Platform', 'Year of Release', 'Genre', 'NA sales', 'EU sales',
       'JP sales', 'Other sales', 'Critic Score', 'User Score', 'Rating'],
      dtype='object')

По результатам проверки видно, что названия столбцов содержат пробелы и используют смешанный регистр символов. Такой формат не является ошибкой, однако может затруднять работу с данными и снижать читаемость кода.

Для приведения датафрейма к единому стилю оформления преобразуем названия столбцов в формат `snake_case`: переведём все символы в нижний регистр и заменим пробелы символом подчёркивания.

In [59]:
# Приводим названия столбцов к стилю snake_case
games.columns = games.columns.str.lower().str.replace(' ', '_')

In [60]:
# Проверяем результат преобразования
games.columns

Index(['name', 'platform', 'year_of_release', 'genre', 'na_sales', 'eu_sales',
       'jp_sales', 'other_sales', 'critic_score', 'user_score', 'rating'],
      dtype='object')

После преобразования к стилю `snake_case` названия столбцов стали более удобными для дальнейшей работы и обращения к ним в коде.

### 2.2. Типы данных

На этапе первичного знакомства с данными было выявлено, что не все типы данных соответствуют содержимому столбцов.

Особое внимание следует уделить столбцам `eu_sales`, `jp_sales` и `user_score`, которые имеют тип данных `object`, хотя содержат числовые показатели. Наиболее вероятной причиной является наличие строковых значений, пропусков или других некорректных записей, препятствующих автоматическому определению числового типа данных.

Начнём со столбца `user_score`. Проверим, присутствуют ли в столбце значения, которые невозможно преобразовать в число и которые могли привести к некорректному определению типа данных.

In [61]:
# Выводим значения столбца user_score, которые невозможно преобразовать в число
games[pd.to_numeric(games['user_score'], errors='coerce').isna()]['user_score'].unique()

array([nan, 'tbd'], dtype=object)

Проверка показала, что помимо пропущенных значений в столбце присутствует значение `tbd` (to be determined). Данная запись используется для обозначения отсутствующей оценки и не может быть интерпретирована как число.

Заменим значение `tbd` на пропуск и преобразуем столбец к числовому типу данных.

In [62]:
# Заменяем значение tbd на пропуск
games['user_score'] = games['user_score'].replace('tbd', None)

In [63]:
# Преобразуем столбец user_score в числовой тип данных
games['user_score'] = pd.to_numeric(games['user_score'])

In [64]:
# Проверяем тип данных столбца user_score
games['user_score'].dtype

dtype('float64')

После замены значения `tbd` и преобразования столбца `user_score` тип данных был успешно изменён на `float64`.

Далее проверим столбцы `eu_sales` и `jp_sales`. Необходимо убедиться, что в данных столбцах отсутствуют нечисловые значения, препятствующие преобразованию к числовому типу данных.

In [65]:
# Выводим значения столбца eu_sales, которые невозможно преобразовать в число
games.loc[pd.to_numeric(games['eu_sales'], errors='coerce').isna(),
    'eu_sales'].unique()

array(['unknown'], dtype=object)

In [66]:
# Выводим значения столбца eu_sales, которые невозможно преобразовать в число
games.loc[pd.to_numeric(games['jp_sales'], errors='coerce').isna(),
    'jp_sales'].unique()

array(['unknown'], dtype=object)

Проверка показала, что в столбцах `eu_sales` и `jp_sales` присутствует значение `unknown`, которое не может быть интерпретировано как число.

Вероятно, данная запись используется для обозначения отсутствующей информации о продажах в соответствующем регионе. Заменим значение `unknown` на пропуск и преобразуем оба столбца к числовому типу данных.

In [67]:
# Заменяем значение unknown на пропуск
games['eu_sales'] = games['eu_sales'].replace('unknown', None)
games['jp_sales'] = games['jp_sales'].replace('unknown', None)

In [16]:
# Преобразуем столбцы продаж к числовому типу данных
games['eu_sales'] = pd.to_numeric(games['eu_sales'])
games['jp_sales'] = pd.to_numeric(games['jp_sales'])

In [17]:
# Проверяем типы данных после преобразования
games[['eu_sales', 'jp_sales']].dtypes

eu_sales    float64
jp_sales    float64
dtype: object

Таким образом, все столбцы, в которых были обнаружены некорректные типы данных, успешно преобразованы.

Столбцы `user_score`, `eu_sales` и `jp_sales` теперь имеют числовой тип данных, что позволит корректно выполнять расчёты и статистический анализ

### 2.3. Наличие пропусков в данных

На предыдущих этапах были исправлены названия столбцов и скорректированы типы данных. Далее необходимо изучить пропущенные значения, определить возможные причины их появления и выбрать способ обработки для каждого случая.

Для этого рассчитаем количество и долю пропусков в каждом столбце датафрейма.

In [18]:
# Выводим количество пропущенных строк в датафрейме
games.isna().sum()

name                  2
platform              0
year_of_release     275
genre                 2
na_sales              0
eu_sales              0
jp_sales              0
other_sales           0
critic_score       8714
user_score         8080
rating             6871
dtype: int64

In [19]:
# Подсчитываем процент строк с пропусками
games.isna().sum() / len(games) * 100

name                0.011795
platform            0.000000
year_of_release     1.621845
genre               0.011795
na_sales            0.000000
eu_sales            0.000000
jp_sales            0.000000
other_sales         0.000000
critic_score       51.391838
user_score         47.652748
rating             40.522529
dtype: float64

После более детального изучения было установлено, что пропущенные значения присутствуют в шести столбцах:

- `name` — 2 пропуска, или 0,01%
- `year_of_release` — 275 пропусков, или 1,62%
- `genre` — 2 пропуска, или 0,01%
- `critic_score` — 8714 пропусков, или 51,39%
- `user_score` — 8080 пропусков, или 47,65%
- `rating` — 6871 пропуск, или 40,52%

Наибольшая доля пропусков наблюдается в столбцах `critic_score`, `user_score` и `rating`. Это может быть связано с тем, что не для всех игр были доступны оценки пользователей и критиков, а также информация о возрастном рейтинге ESRB.

Пропуски в столбцах `name` и `genre` составляют крайне небольшую долю данных. Вероятнее всего, они появились в результате ошибок при заполнении или сборе информации.

Пропуски в столбце `year_of_release` встречаются относительно редко, однако данный столбец играет важную роль в дальнейшем анализе, поскольку на его основе будет формироваться временной срез данных.

Таким образом, наибольшее внимание следует уделить столбцам с оценками и рейтингом, а также определить способ обработки пропусков в году выпуска игр.

Начнём с пропусков в столбцах `name` и `genre`.

Поскольку количество таких записей крайне мало, необходимо проверить, содержат ли строки дополнительную информацию, которая позволит восстановить пропущенные значения. Если восстановить данные невозможно, такие записи можно удалить без существенного влияния на результаты анализа.


In [20]:
# Выводим строки с пропусками в названии игры или жанре
games[games['name'].isna() | games['genre'].isna()]

,name,platform,year_of_release,genre,na_sales,eu_sales,jp_sales,other_sales,critic_score,user_score,rating
661,NaN,GEN,1993.0,NaN,1.78,0.53,0.00,0.08,NaN,NaN,NaN
14439,NaN,GEN,1993.0,NaN,0.00,0.00,0.03,0.00,NaN,NaN,NaN


Проверка показала, что обе записи с пропущенными значениями содержат одновременно пропуски в столбцах `name` и `genre`.

Кроме того, данные строки не содержат пользовательских и экспертных оценок, а также возрастного рейтинга. Восстановить название игры или её жанр по имеющейся информации не представляется возможным.

Поскольку количество таких записей составляет всего 2 строки (около 0,01% от общего объёма данных), их удаление не окажет существенного влияния на результаты исследования. Поэтому удалим данные записи из датафрейма.

In [21]:
# Удаляем строки с пропусками в названии игры и жанре
games = games.dropna(subset=['name', 'genre'])

In [22]:
# Проверяем результат удаления
games[games['name'].isna() | games['genre'].isna()]

,name,platform,year_of_release,genre,na_sales,eu_sales,jp_sales,other_sales,critic_score,user_score,rating


Далее рассмотрим пропуски в столбце `year_of_release`.

Данный столбец играет важную роль в исследовании, поскольку на его основе будет формироваться выборка игр за период с 2000 по 2013 год. Поэтому необходимо определить возможность восстановления пропущенных значений или выбрать другой способ их обработки.

In [23]:
# Выводим строки с пропусками в году выпуска
games[games['year_of_release'].isna()]

,name,platform,year_of_release,genre,na_sales,eu_sales,jp_sales,other_sales,critic_score,user_score,rating
183,Madden NFL 2004,PS2,NaN,Sports,4.26,0.26,0.01,0.71,94.0,8.5,E
379,FIFA Soccer 2004,PS2,NaN,Sports,0.59,2.36,0.04,0.51,84.0,6.4,E
458,LEGO Batman: The Videogame,Wii,NaN,Action,1.80,0.97,0.00,0.29,74.0,7.9,E10+
477,wwe Smackdown vs. Raw 2006,PS2,NaN,Fighting,1.57,1.02,0.00,0.41,NaN,NaN,NaN
611,Space Invaders,2600,NaN,Shooter,2.36,0.14,0.00,0.03,NaN,NaN,NaN
...,...,...,...,...,...,...,...,...,...,...,...
16609,PDC World Championship Darts 2008,PSP,NaN,Sports,0.01,0.00,0.00,0.00,43.0,NaN,E10+
16641,Freaky Flyers,GC,NaN,Racing,0.01,0.00,0.00,0.00,69.0,6.5,T
16685,Inversion,PC,NaN,Shooter,0.01,0.00,0.00,0.00,59.0,6.7,M
16695,Hakuouki: Shinsengumi Kitan,PS3,NaN,Adventure,0.01,0.00,0.00,0.00,NaN,NaN,NaN


При просмотре строк с пропущенным годом выпуска видно, что часть игр содержит другую полезную информацию: название, платформу, жанр, продажи, оценки и рейтинг.

Однако восстановить год выпуска автоматически по имеющимся данным нельзя. Название игры не всегда однозначно указывает на год выхода, а одна и та же игра могла выпускаться на разных платформах в разные годы.

Так как столбец `year_of_release` необходим для дальнейшей фильтрации данных по периоду с 2000 по 2013 год, строки без года выпуска не подходят для дальнейшего анализа. Доля таких строк составляет 1,62%, поэтому их удаление не должно существенно повлиять на результат исследования.

In [24]:
# Удаляем строки с пропусками в году выпуска
games = games.dropna(subset=['year_of_release'])

In [25]:
# Преобразуем столбец year_of_release в целочисленный тип данных
games['year_of_release'] = games['year_of_release'].astype('int')

In [26]:
# Проверяем тип данных столбца year_of_release
games['year_of_release'].dtype

dtype('int64')

In [27]:
# Проверяем, остались ли пропуски в ключевых столбцах
games[['name', 'genre', 'year_of_release']].isna().sum()

name               0
genre              0
year_of_release    0
dtype: int64

Рассмотрим пропуски в столбцах `critic_score`, `user_score` и `rating`.

На данные столбцы приходится наибольшее количество пропущенных значений в датафрейме. Все они содержат информацию об оценках игры или её возрастном рейтинге.

Поскольку отсутствие оценки или рейтинга само по себе не является ошибкой в данных, важно определить, целесообразно ли заполнять такие пропуски искусственными значениями или лучше сохранить их в исходном виде.

In [28]:
# Выводим количество пропусков в столбцах с оценками и рейтингом
games[['critic_score', 'user_score', 'rating']].isna().sum()

critic_score    8594
user_score      7957
rating          6778
dtype: int64

Пропуски в столбцах `critic_score` и `user_score`, вероятно, связаны с отсутствием достаточного количества обзоров критиков или пользовательских оценок для отдельных игр.

Заполнять данные пропуски средними или медианными значениями нецелесообразно, поскольку это может привести к искажению результатов анализа. Отсутствие оценки не означает, что игра имеет среднюю или какую-либо другую оценку.

Поэтому пропуски в столбцах `critic_score` и `user_score` будут сохранены без изменений.

Рассмотрим отдельно столбец `rating`. Поскольку данный признак является категориальным, вместо удаления большого количества записей можно использовать специальное значение-индикатор. Предварительно проверим, отсутствует ли выбранное значение среди существующих категорий рейтинга ESRB.

In [29]:
# Выводим уникальные значения столбца rating
games['rating'].dropna().unique()

array(['E', 'M', 'T', 'E10+', 'K-A', 'AO', 'EC', 'RP'], dtype=object)

In [30]:
# Заменяем пропуски в рейтинге на значение-индикатор
games['rating'] = games['rating'].fillna('UNKNOWN')

In [31]:
# Проверяем количество пропусков после замены
games['rating'].isna().sum()

0

Проверка показала, что значение `UNKNOWN` отсутствует среди существующих категорий рейтинга ESRB, поэтому его можно использовать в качестве значения-индикатора.

Пропуски в столбце `rating` были заменены на значение `UNKNOWN`, которое обозначает отсутствие информации о возрастном рейтинге игры и не пересекается с существующими категориями ESRB.

Таким образом, все пропуски в столбце `rating` были обработаны, а пропуски в столбцах `critic_score` и `user_score` сохранены без изменений.

### 2.4. Явные и неявные дубликаты в данных

После обработки пропущенных значений необходимо проверить данные на наличие явных и неявных дубликатов.

Неявные дубликаты могут возникать из-за различий в регистре символов, опечаток или различных способов написания одних и тех же значений. Особое внимание следует уделить категориальным столбцам, таким как название платформы, жанр игры и возрастной рейтинг.

После проверки и устранения возможных неявных дубликатов необходимо проверить датафрейм на наличие полностью совпадающих строк.

И при необходимости провести нормализацию данных с текстовыми значениями. Названия жанров игр привести к нижнему регистру, а названия рейтинга - к верхнему.

In [32]:
# Выводим уникальные значения платформ
sorted(games['platform'].unique())

['2600',
 '3DO',
 '3DS',
 'DC',
 'DS',
 'GB',
 'GBA',
 'GC',
 'GEN',
 'GG',
 'N64',
 'NES',
 'NG',
 'PC',
 'PCFX',
 'PS',
 'PS2',
 'PS3',
 'PS4',
 'PSP',
 'PSV',
 'SAT',
 'SCD',
 'SNES',
 'TG16',
 'WS',
 'Wii',
 'WiiU',
 'X360',
 'XB',
 'XOne']

In [33]:
# Выводим уникальные значения жанров
sorted(games['genre'].unique())

['ACTION',
 'ADVENTURE',
 'Action',
 'Adventure',
 'FIGHTING',
 'Fighting',
 'MISC',
 'Misc',
 'PLATFORM',
 'PUZZLE',
 'Platform',
 'Puzzle',
 'RACING',
 'ROLE-PLAYING',
 'Racing',
 'Role-Playing',
 'SHOOTER',
 'SIMULATION',
 'SPORTS',
 'STRATEGY',
 'Shooter',
 'Simulation',
 'Sports',
 'Strategy']

In [34]:
# Выводим уникальные значения рейтингов
sorted(games['rating'].unique())

['AO', 'E', 'E10+', 'EC', 'K-A', 'M', 'RP', 'T', 'UNKNOWN']

Проверка уникальных значений показала, что в столбце `platform` неявные дубликаты отсутствуют. Названия платформ записаны единообразно.

В столбце `genre` были обнаружены неявные дубликаты, связанные с различием регистра символов. Например, один и тот же жанр представлен значениями `ACTION` и `Action`, `SPORTS` и `Sports`, `RACING` и `Racing`.

Для устранения подобных дубликатов приведём все значения столбца `genre` к единому формату записи.

Также при анализе уникальных значений столбца `rating` было обнаружено, что категории `K-A` и `E` фактически обозначают одну и ту же возрастную группу.

Рейтинг `K-A` (Kids to Adults) использовался организацией ESRB до 1998 года и впоследствии был заменён на рейтинг `E` (Everyone). Поскольку данные категории являются эквивалентными, для обеспечения единообразия данных заменим значение `K-A` на `E`.

In [35]:
# Приводим названия жанров к единому регистру
games['genre'] = games['genre'].str.lower()

In [36]:
# Проверяем результат преобразования
sorted(games['genre'].unique())

['action',
 'adventure',
 'fighting',
 'misc',
 'platform',
 'puzzle',
 'racing',
 'role-playing',
 'shooter',
 'simulation',
 'sports',
 'strategy']

In [37]:
# Заменяем устаревший рейтинг K-A на E
games['rating'] = games['rating'].replace('K-A', 'E')

In [38]:
# Проверяем уникальные значения рейтингов
sorted(games['rating'].unique())

['AO', 'E', 'E10+', 'EC', 'M', 'RP', 'T', 'UNKNOWN']

После обработки неявных дубликатов в столбцах `genre` и `rating` необходимо проверить, есть ли в датафрейме полностью совпадающие строки.

Такие строки считаются явными дубликатами. Если они присутствуют, их нужно удалить, так как повторяющиеся записи могут исказить результаты дальнейшего анализа.

In [39]:
# Подсчитываем количество явных дубликатов
games.duplicated().sum()

235

Проверка показала наличие 235 явных дубликатов.

Поскольку такие записи полностью повторяют друг друга и не несут новой информации, их необходимо удалить. Сохранение дубликатов может привести к искажению результатов анализа и завышению отдельных показателей.

Удалим найденные дубликаты и повторно проверим датафрейм.

In [40]:
# Удаляем явные дубликаты
games = games.drop_duplicates()

In [41]:
# Проверяем результат удаления
games.duplicated().sum()

0

In [42]:
# Определяем количество удалённых строк
16956 - len(games)

512

In [43]:
# Определяем долю удалённых строк
(16956 - len(games)) / 16956 * 100

3.019580089643784

В процессе предобработки данных было удалено 512 строк, что составляет около 3,02% от исходного объёма датасета.

Удалённые записи включают строки с отсутствующими значениями в критически важных столбцах (`name`, `genre`, `year_of_release`), а также обнаруженные явные дубликаты. Доля удалённых данных является небольшой, поэтому данная операция не должна оказать существенного влияния на результаты дальнейшего анализа.

#### Общий вывод по предобработке данных

В ходе предобработки были выполнены основные этапы очистки и подготовки данных к анализу.

Названия столбцов были приведены к стилю `snake_case`, что повысило удобство работы с датафреймом. Также были исправлены некорректные типы данных в столбцах `user_score`, `eu_sales` и `jp_sales`. Столбец `year_of_release` после удаления пропусков был преобразован к целочисленному типу данных.

Был проведён анализ пропущенных значений. Строки с отсутствующими значениями в столбцах `name`, `genre` и `year_of_release` были удалены, поскольку восстановить данные значения не представлялось возможным или они были необходимы для дальнейшего анализа. Пропуски в столбцах `critic_score` и `user_score` были сохранены, так как отражают отсутствие информации об оценках игр. Для столбца `rating` пропуски были заменены на значение-индикатор `UNKNOWN`.

В ходе поиска неявных дубликатов были приведены к единому формату записи значения столбцов `name` и `genre`, а устаревшая категория рейтинга `K-A` была заменена на её современный аналог `E`. Кроме того, было обнаружено и удалено 235 явных дубликатов.

Всего в процессе предобработки было удалено 512 строк, что составляет около 3,02% исходного объёма данных. После выполнения всех преобразований данные готовы к дальнейшему анализу.

---

## 3. Фильтрация данных

Для дальнейшего анализа необходимо рассмотреть игры, выпущенные в период с 2000 по 2013 год включительно.

Сформируем отдельный датафрейм `df_actual`, содержащий только записи, относящиеся к указанному временному периоду.

In [44]:
# Формируем срез данных за период с 2000 по 2013 год включительно
df_actual = games[(games['year_of_release'] >= 2000) &
    (games['year_of_release'] <= 2013)]

In [45]:
# Выводим размер полученного датафрейма
df_actual.shape

(12781, 11)

После фильтрации в датафрейме осталось 12781 запись и 11 столбцов. Полученный срез будет использоваться на следующих этапах исследования для категоризации данных и анализа платформ.

Таким образом, сформирована актуальная выборка, отражающая историю продаж компьютерных игр в начале XXI века.

---

## 4. Категоризация данных

После формирования необходимого временного среза выполним категоризацию игр по пользовательским и экспертным оценкам.

Пользовательские оценки будут разделены на следующие категории:

- низкая оценка - от 0 до 3, не включая правую границу интервала
- средняя оценка - от 3 до 8, не включая правую границу интервала
- высокая оценка - от 8 до 10 включительно

Оценки критиков будут разделены аналогичным образом:

- низкая оценка - от 0 до 30, не включая правую границу интервала
- средняя оценка - от 30 до 80, не включая правую границу интервала
- высокая оценка - от 80 до 100 включительно

Для удобства анализа будут созданы отдельные столбцы с категориями пользовательских и экспертных оценок.
Кроме того, будет определён топ-7 платформ, на которых за рассматриваемый период было выпущено наибольшее количество игр.

Начнём с пользовательских оценок. Для каждой игры определим категорию оценки в соответствии с указанными интервалами.

In [46]:
# Создаем функцию категоризации пользовательских оценок
def user_score_category(score):
    if pd.isna(score):
        return 'нет оценки'
    elif 0 <= score < 3:
        return 'низкая'
    elif 3 <= score < 8:
        return 'средняя'
    elif 8 <= score <= 10:
        return 'высокая'

In [47]:
# Создаем новый столбец с категориями пользовательских оценок
df_actual['user_score_category'] = df_actual['user_score'].apply(user_score_category)

/tmp/ipykernel_92/2639437974.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_actual['user_score_category'] = df_actual['user_score'].apply(user_score_category)


In [48]:
# Подсчитываем количество игр в каждой категории пользовательских оценок
df_actual['user_score_category'].value_counts()

нет оценки    5195
средняя       4820
высокая       2621
низкая         145
Name: user_score_category, dtype: int64

Пользовательские оценки разделены на категории в соответствии с условиями задания.

Наибольшее количество игр относится к категории без пользовательской оценки - 5195 записей. Это подтверждает вывод, сделанный ранее при анализе пропусков: значительная часть игр не имеет пользовательских оценок.

Среди игр, для которых оценки доступны, преобладает категория со средней оценкой - 4820 игр. Высокую оценку получили 2621 игра, тогда как низкая оценка встречается значительно реже - всего у 145 игр.

Таким образом, большинство игр с доступными пользовательскими оценками имеют средний или высокий уровень оценки.

Аналогичным образом выполним категоризацию оценок критиков.
Для определения категорий будут использоваться интервалы, указанные ранее.

In [49]:
# Создаем функцию категоризации оценок критиков
def critic_score_category(score):
    if pd.isna(score):
        return 'нет оценки'
    elif 0 <= score < 30:
        return 'низкая'
    elif 30 <= score < 80:
        return 'средняя'
    elif 80 <= score <= 100:
        return 'высокая'

In [50]:
# Создаем новый столбец с категориями оценок критиков
df_actual['critic_score_category'] = df_actual['critic_score'].apply(critic_score_category)

/tmp/ipykernel_92/1896175276.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_actual['critic_score_category'] = df_actual['critic_score'].apply(critic_score_category)


In [51]:
# Подсчитываем количество игр в каждой категории оценок критиков
df_actual['critic_score_category'].value_counts()

нет оценки    5612
средняя       5422
высокая       1692
низкая          55
Name: critic_score_category, dtype: int64

Оценки критиков также успешно разделены на категории в соответствии с заданными интервалами.

Наибольшее количество игр относится к категории без оценки критиков - 5612 записей. Это соответствует результатам анализа пропусков, выполненного на этапе предобработки данных.

Среди игр с доступными оценками критиков преобладает категория со средней оценкой - 5422 игры. Высокую оценку получили 1692 игры, тогда как низкие оценки встречаются достаточно редко - всего у 55 игр.

Таким образом, как и в случае с пользовательскими оценками, большинство игр с доступными оценками критиков относятся к категории средних или высоких оценок.

Завершающим этапом категоризации выделим платформы, на которых за рассматриваемый период было выпущено наибольшее количество игр.

Для этого сгруппируем данные по платформам и определим семь платформ с максимальным количеством выпущенных игр.

In [52]:
# Определяем топ-7 платформ по количеству выпущенных игр
top_platforms = (df_actual['platform'].value_counts().head(7))
top_platforms

PS2     2127
DS      2120
Wii     1275
PSP     1180
X360    1121
PS3     1087
GBA      811
Name: platform, dtype: int64

По количеству выпущенных игр за период с 2000 по 2013 год лидирует платформа `PS2`, для которой было выпущено 2127 игр. Практически такой же результат показывает платформа `DS` - 2120 игр.

В топ-7 также вошли платформы `Wii`, `PSP`, `X360`, `PS3` и `GBA`. Количество выпущенных игр на этих платформах варьируется от 811 до 1275.

Полученные результаты показывают, что в рассматриваемый период наибольшее количество игр выпускалось преимущественно для консолей компаний Sony и Nintendo.

---

## 5. Итоговый вывод

В ходе выполнения проекта были изучены данные о продажах компьютерных игр на различных платформах, а также информация об их жанрах, пользовательских и экспертных оценках и возрастных рейтингах.

На этапе предобработки были выполнены следующие действия:
- названия столбцов приведены к стилю `snake_case`
- исправлены некорректные типы данных в столбцах `user_score`, `eu_sales` и `jp_sales`
- обработаны пропущенные значения
- удалены строки с отсутствующими значениями в столбцах `name`, `genre` и `year_of_release`
- пропуски в столбце `rating` заменены на значение-индикатор `UNKNOWN`
- устранены неявные дубликаты в столбцах `genre` и `rating`
- удалены 235 явных дубликатов

Всего в процессе предобработки было удалено 512 строк, что составляет около 3,02% исходного объёма данных.

После очистки данных был сформирован отдельный срез `df_actual`, содержащий только игры, выпущенные в период с 2000 по 2013 год включительно. После фильтрации в выборке осталось 12781 запись, что позволило сосредоточиться на исследовании рынка компьютерных игр начала XXI века.

Для выполнения категоризации в датафрейм были добавлены два новых поля:
- `user_score_category` - категория пользовательской оценки (`низкая`, `средняя`, `высокая`, `нет оценки`);
- `critic_score_category` - категория оценки критиков (`низкая`, `средняя`, `высокая`, `нет оценки`).

Категоризация показала, что среди игр с доступными оценками преобладают средние и высокие оценки как со стороны пользователей, так и со стороны критиков. Количество игр с низкими оценками сравнительно невелико.

Также был определён топ-7 платформ по количеству выпущенных игр за рассматриваемый период. Лидерами стали платформы `PS2` и `DS`, для которых было выпущено более двух тысяч игр.